onnxruntime/python/tools/quantization/quant_utils.py

```diff
@@ -482,6 +466,36 @@ def normalize_axis(axis: int, rank: int) -> tuple[bool, int]:
     return is_valid, axis_norm
 
 
+def pack_bytes_to_4bit(src_8bit: bytes) -> bytearray:
+    """
+    Copies a source array of 8-bit values into a destination bytearray of packed 4-bit values.
+    Assumes that the source values are already in the appropriate int4 range.
+    :parameter src_8bit: The 8-bit element values to pack.
+    :return A bytearray with every two 8-bit src elements packed into a single byte.
+    """
+    num_elems = len(src_8bit)
+    if num_elems == 0:
+        return bytearray()
+
+    dst_size = (num_elems + 1) // 2  # Ex: 5 8-bit elems packed into 3 bytes
+    dst = bytearray(dst_size)
+
+    src_i: int = 0
+    dst_i: int = 0
+
+    # Pack two 8-bit elements into a single byte in each iteration.
+    while src_i < num_elems - 1:
+        dst[dst_i] = ((src_8bit[src_i + 1] & 0xF) << 4) | (src_8bit[src_i] & 0xF)
+        dst_i += 1
+        src_i += 2
```

Comment:
just for my curiosity: is it possible to use np strided slice? like, `dst[0:] |= src_8bit[0::2], dst[0:] |= src_8bit[1::2] << 4`

Reply:
That's a good question! I'm afraid I'm not too experienced with np strided slices. Unfortunately, I'm not sure if this would be faster without more profiling. Also, this code (as is) throws a TypeError and does not discard the top 4bits of the elements in `src`. Maybe we can explore this separately?

Reply:
definitely! thanks for your input